### ISSUE TITLE
RRTMGP and Single Precision

### ISSUE THREAD

**Issue by mathomp4** [earth-system-radiation/rte-rrtmgp#156]
TL;DR: Is there an "official" way to run RRTMGP in single-precision?

---

So this is a bit of a long ramble of a question so bear with me...

In trying to get RRTMGP v1.5 working in GEOS, we found it was crashing with a segfault...and the previous version didn't seem to the last we tested. 

Eventually, I was able to figure out that, for some reason, the set of Intel Fortran flags we switched to recently--which enable vectorization model-wide--lead to a segfault in RRTMGP, and we made that switch in between tests. Now, I don't think the actual issue is in RRTMGP since changing the flags back to our old non-vectorized flags lets it work. I've tried to track down which of our flags cause the issue, but no luck so far.

But I got curious and thought, "I wonder if I can run RRTMGP in single-precision?" for some extra speed (since part of my job is code optimization). So I changed `mo_rte_kind.F90`, recompiled, and boom. Big crash! But, a bit of GitHub searching led me to [this commit](https://github.com/peterukk/rte-rrtmgp-nn/commit/abba4a9a485041babfb732bec954d62acca9dde0) on the rte-rrtgmp-nn repo of @peterukk. So, I added a `k_min`, made similar edits, and I can run RRTMGP in single precision! Even weirder, it works with *both* non-vectorized and vectorized flags (which *really* makes me think someone else is messing with RRTMGP's memory state.)

And this got me thinking: is there a "correct" way to run single-precision RRTMGP? My naive attempt at just changing `wp` failed, but with the tweak from @peterukk it did seem to work and the output wasn't a different planet...

**Comment by Chiil** [comment]
At the moment it is unsupported, but @peterukk and we got it working in SP in the C++ code. There is some pretty recent discussion here, @peterukk has a few fixes that could be of interest for your: https://github.com/earth-system-radiation/rte-rrtmgp/pull/39

**Comment by mathomp4** [comment]
> At the moment it is unsupported, but @peterukk has some fixes for the Fortran code, and we got it working in SP in the C++ code. There is some pretty recent discussion here, @peterukk has a few fixes that could be of interest for you: #39

Well I feel dumb. Comments from yesterday about this! 😄 

Probably I never saw it because I don't know the right words to search for. I kept looking for `REAL32` or the like. (Can you tell I'm not an atmospheric scientist?)

**Comment by peterukk** [comment]
Not at all, it was quite hidden in the pull request topics. Would be great if more people could test the fix! Although it sounds like Robert will put it in the development branch soon anyways 

Again, credit for the clever stuff with the physical check goes to someone else :smile:

**Comment by mathomp4** [comment]
Query: We (well, @dr0cloud) are looking at integrating RRTMGP v1.6 in GEOS and I was wondering: is SP RRTMGP now in 1.6? It's been about a year since I made this question/issue so I figure if it is in now, I can close it.

**Comment by RobertPincus** [comment]
@mathomp4 Egg on my face - this took a back seat (including to teaching) but is coming this Spring... FWIW the issue is less the actual coding as much as implementing single precision in the automated testing.

**Comment by mathomp4** [comment]
> @mathomp4 Egg on my face - this took a back seat (including to teaching) but is coming this Spring... FWIW the issue is less the actual coding as much as implementing single precision in the automated testing.

@RobertPincus No rush. I figured you'd had one or two versions of RRTMGP since I made this issue and figured I'd ask.

And I should probably figure out if we still need to change compiler flags for RRTMGP. I mean, it's been a year and most likely it was a compiler bug! 😄

**Comment by peterukk** [comment]
FYI I think the physical check in the shortwave reflectance-transmittance kernel is outdated in my repo, it should have a factor mu0 (copypasting ecRad code, I hope the variable names are clear):

```
    ! Final check that ref_dir + trans_dir_diff <= 1
      ref_dir(jg)        = max(0.0_jprb, min(ref_dir(jg), mu0*(1.0_jprb-trans_dir_dir(jg))))
      trans_dir_diff(jg) = max(0.0_jprb, min(trans_dir_diff(jg), mu0*(1.0_jprb-trans_dir_dir(jg))-ref_dir(jg)))
```

We've done many IFS experiments with radiation fully in single precision, including longer climate runs, never ran into crashes or other problems.

**Comment by RobertPincus** [comment]
@peterukk Thanks for this pointer. What's trans_dir_dir? Is it the transmissivity for the direct beam `exp(-tau/mu0)`?

**Comment by peterukk** [comment]
> @peterukk Thanks for this pointer. What's trans_dir_dir? Is it the transmissivity for the direct beam `exp(-tau/mu0)`?

Yes, that's correct

**Comment by peterukk** [comment]
RTE version: 

```
        ! Final check that energy is not spuriously created, by recognizing that
        ! the beam can either be reflected, penetrate unscattered to the base of a layer, 
        ! or penetrate through but be scattered on the way
        ! Makes the equations safer in single precision. Credit: Robin Hogan
        Rdir    = max(0.0_wp, min(Rdir, (1.0_wp - Tnoscat(igpt)       ) ))
        Tdir    = max(0.0_wp, min(Tdir, (1.0_wp - Tnoscat(igpt) - Rdir) ))

        source_up  (igpt,ilev)  =   Rdir    *   dir_flux_inc(igpt) 
        source_dn  (igpt,ilev)  =   Tdir    *   dir_flux_inc(igpt)
        dir_flux_trans(igpt)    =   Tnoscat(igpt) * dir_flux_inc(igpt)
end do
```

**Comment by RobertPincus** [comment]
Closed with 3ac0636
